### ISSUE TITLE
Installing fiona on Linux venv creates issue with libgeos shared library

### ISSUE THREAD

**Issue by mhearne-usgs** [conda-forge/fiona-feedstock#66]
Platform:
uname -a
Linux igskcicgvmlossd.cr.usgs.gov 3.10.0-693.11.1.el7.x86_64 #1 SMP Fri Dec 1 09:09:59 CST 2017 x86_64 x86_64 x86_64 GNU/Linux

Setup:
```shell
conda create --name testme python=3.5
source activate testme
conda install fiona
```

Generating the error:
```shell
python
>>> import osgeo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mhearne/miniconda/envs/testme/lib/python3.5/site-packages/osgeo/__init__.py", line 21, in <module>
    _gdal = swig_import_helper()
  File "/home/mhearne/miniconda/envs/testme/lib/python3.5/site-packages/osgeo/__init__.py", line 17, in swig_import_helper
    _mod = imp.load_module('_gdal', fp, pathname, description)
  File "/home/mhearne/miniconda/envs/testme/lib/python3.5/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/home/mhearne/miniconda/envs/testme/lib/python3.5/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: libgeos-3.5.0.so: cannot open shared object file: No such file or directory
```

**Comment by ocefpaf** [comment]
@mhearne-usgs I tried to reproduce that and I get a bunch of packages that are mixed with `defaults`. Unfortunately `conda` channel preference is an illusion that broken badly when continuum started to build packages with conda-build 3 and update packages constantly based on conda-forge recipes.

For example, here is what I get from your example above:

```
    click:          6.7-py_1              conda-forge
    click-plugins:  1.0.3-py35_0          conda-forge
    cligj:          0.4.0-py35_0          conda-forge
    curl:           7.55.1-0              conda-forge
    fiona:          1.7.11-py35_0         conda-forge
    gdal:           2.1.0-py35_0          defaults   
    geos:           3.5.1-1               conda-forge
    geotiff:        1.4.1-0               defaults   
    hdf4:           4.2.12-0              conda-forge
    hdf5:           1.8.17-11             conda-forge
    icu:            58.2-0                conda-forge
    intel-openmp:   2018.0.0-hc7b2577_8   defaults   
    jpeg:           9b-2                  conda-forge
    kealib:         1.4.6-3               conda-forge
    krb5:           1.14.2-0              conda-forge
    libgcc-ng:      7.2.0-h7cc24e2_2      defaults   
    libgdal:        2.1.0-0               defaults   
    libgfortran:    3.0.0-1               defaults   
    libgfortran-ng: 7.2.0-h9f7466a_2      defaults   
    libnetcdf:      4.4.1-0               conda-forge
    libssh2:        1.8.0-2               conda-forge
    libtiff:        4.0.9-0               conda-forge
    mkl:            2018.0.1-h19d6760_4   defaults   
    munch:          2.2.0-py35_0          conda-forge
    numpy:          1.14.0-py35h3dfced4_0 defaults   
    proj4:          4.9.2-0               defaults   
    shapely:        1.6.2-py35_0          conda-forge
    six:            1.11.0-py35_1         conda-forge
    xerces-c:       3.1.4-3               conda-forge
```

We got `libgdal`, `proj4`, `gdal`, and `geotiff` from `defaults` but fiona is from `conda-forge` which is enough to render `fiona` useless b/c it links to `gdal`.

Note that this does not happen on Python 3.6, which means that **maybe** a package for PY35 is missing or has a bad pinned version lingering in the channel. However, looking at the logs for that [`1.7.11-py35_0`](https://circleci.com/gh/conda-forge/fiona-feedstock/105) everything was fine at the time :unamused: 

```shell
    blas:            1.1-openblas                 conda-forge
    boost:           1.65.1-py35_0                conda-forge
    boost-cpp:       1.65.1-1                     conda-forge
    bzip2:           1.0.6-1                      conda-forge
    ca-certificates: 2017.11.5-0                  conda-forge
    cairo:           1.14.6-5                     conda-forge
    certifi:         2017.11.5-py35_0             conda-forge
    curl:            7.55.1-0                     conda-forge
    cython:          0.27.3-py35_0                conda-forge
    expat:           2.1.0-3                      conda-forge
    fontconfig:      2.12.1-6                     conda-forge
    freetype:        2.7-2                        conda-forge
    freexl:          1.0.4-0                      conda-forge
    gdal:            2.1.4-py35_2                 conda-forge
    geos:            3.6.2-1                      conda-forge
    geotiff:         1.4.2-0                      conda-forge
    gettext:         0.19.7-1                     conda-forge
    giflib:          5.1.4-0                      conda-forge
    glib:            2.51.4-0                     conda-forge
    hdf4:            4.2.13-0                     conda-forge
    hdf5:            1.10.1-1                     conda-forge
    icu:             58.2-0                       conda-forge
    jpeg:            9b-2                         conda-forge
    json-c:          0.12.1-0                     conda-forge
    kealib:          1.4.7-4                      conda-forge
    krb5:            1.14.2-0                     conda-forge
    libdap4:         3.18.3-2                     conda-forge
    libffi:          3.2.1-3                      conda-forge
    libgdal:         2.1.4-8                      conda-forge
    libgfortran:     3.0.0-1                      defaults   
    libiconv:        1.15-0                       conda-forge
    libkml:          1.3.0-4                      conda-forge
    libnetcdf:       4.4.1.1-10                   conda-forge
    libpng:          1.6.28-2                     conda-forge
    libpq:           9.6.3-0                      conda-forge
    libspatialite:   4.3.0a-18                    conda-forge
    libssh2:         1.8.0-2                      conda-forge
    libtiff:         4.0.7-1                      conda-forge
    libxml2:         2.9.5-2                      conda-forge
    ncurses:         5.9-10                       conda-forge
    numpy:           1.8.2-py35_blas_openblas_201 conda-forge [blas_openblas]
    openblas:        0.2.20-6                     conda-forge
    openjpeg:        2.1.2-3                      conda-forge
    openssl:         1.0.2n-0                     conda-forge
    pcre:            8.39-0                       conda-forge
    pixman:          0.34.0-1                     conda-forge
    poppler:         0.61.1-1                     conda-forge
    poppler-data:    0.4.8-0                      conda-forge
    proj4:           4.9.3-5                      conda-forge
    python:          3.5.4-2                      conda-forge
    readline:        6.2-0                        conda-forge
    setuptools:      38.2.4-py35_0                conda-forge
    sqlite:          3.13.0-1                     conda-forge
    tk:              8.5.19-2                     conda-forge
    util-linux:      2.21-0                       defaults   
    xerces-c:        3.2.0-0                      conda-forge
    xz:              5.2.3-0                      conda-forge
    zlib:            1.2.11-0                     conda-forge
```



TL;DR I'll bump the build number and see how it goes, maybe a new build will solve things or help us find the problem.

Ping @msarahan and @pelson just to provide another data point of how bad things are getting. The more we delay to move to conda-build 3 the worst they'll get.

**Comment by msarahan** [comment]
whoa, this is a train wreck.  Unfortunately, I don't think it's as simple as channel priority not working.  Specifically, you have some really old stuff coming in from defaults:

```
    gdal:           2.1.0-py35_0          defaults
    proj4:          4.9.2-0               defaults 
    libgdal:        2.1.0-0               defaults   
    libgfortran:    3.0.0-1               defaults
```

That's from the "free" channel.  I think the major problem is that the stuff on the free channel wasn't pinned intelligently (or at all), and thus it is appearing sometimes when the newer, more intelligently pinned stuff (built with cb3) is either unsatisfiable or less desirable somehow.

Unfortunately, the complicated web of dependencies here makes that extraordinarily difficult to identify.  I'll ask @mcg1969 if he might have any tips for pulling this apart and understanding it.

**Comment by ocefpaf** [comment]
> Unfortunately, I don't think it's as simple as channel priority not working. 

I understand your point of view there but IMO it is b/c channel priority should not let conda look for packages in another unless it is really unavailable in the first one. Anyways, we do have a similar situation now with `libkml 1.3.0`  from `defaults` breaking `conda-forge` `gdal` b/c `libkml 1.3.0` from conda-forge uses `boost` at run time while defaults uses `libboost` instead. Again, an issue that only channel preference can solve b/c dependency solvers will never know what to do when a dependency name is different.

**Comment by mcg1969** [comment]
> channel priority should not let conda look for packages in another unless it is really unavailable in the first one

That's an interesting way of looking at channel priority. I don't think I really believe that anyone would want that, even if they _think_ they do.

I'll see if I can figure out which package is forcing conda to reach down into defaults.

**Comment by ocefpaf** [comment]
>  I don't think I really believe that anyone would want that, even if they _think_ they do.

Why? Everything using OpenSUSE Linux repositories priorities get that feature and it works wonders. Not to mention that many hardcore `conda-forge` users disable `defaults` entirely just to get away from bad conda solver solutions like that one. Anyway, I am not interest in prolonging this discussion, specially ever since `conda-build 3` was released and the new features seems enough to solve issues like this one.

**Comment by mcg1969** [comment]
I'm not sure I'd consider the "nearly-all-conda-forge" recipe to be "fine", by the way. It reaches all the way back to NumPy 1.8, _and_ it still reaches into `defaults` for two packages.

**Comment by mcg1969** [comment]
> Not to mention that many hardcore conda-forge users disable defaults entirely just to get away from bad conda solver solutions like that one.

Yes, they do that, until they can't. This recipe, for instance, cannot be built that way.

**Comment by mcg1969** [comment]
```
PackageNotFoundError: Packages missing in current channels:
            
  - fiona -> gdal 2.1.* -> libgdal 2.1.4 -> hdf5 1.10.1 -> libgfortran
  - fiona -> gdal 2.1.* -> libgdal 2.1.4 -> kealib 1.4.7
  - fiona -> gdal 2.1.* -> libgdal 2.1.4 -> libkml 1.3.0 -> boost 1.66.0 -> numpy >=1.8 -> blas 1.1 openblas -> openblas
  - fiona -> gdal 2.1.* -> libgdal 2.1.4 -> libkml 1.3.0 -> boost 1.66.0 -> numpy >=1.8 -> @blas_openblas
  - fiona -> shapely

We have searched for the packages in the following channels:
            
  - https://conda.anaconda.org/conda-forge/linux-64
  - https://conda.anaconda.org/conda-forge/noarch
```

**Comment by mcg1969** [comment]
It looks like `libgfortran` is the primary culprit (`shapely` depends on that, too)

**Comment by mcg1969** [comment]
One more thing about the channel priority issue---conda 4.4 makes it far easier to second-guess the solver, by requiring individual package to come from particular channels.

**Comment by jorisvandenbossche** [comment]
FYI, this failure is also present on python 2.7

**Comment by ocefpaf** [comment]
Thanks for the report @jorisvandenbossche. I'm waiting on Travis-CI to start a "chain of merges" that should solve this.
